// passport-steam-openid/dist/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VALID_SIGNED_FIELD = exports.VALID_ASSOC_HANDLE = exports.PLAYER_SUMMARY_URL = exports.VALID_OPENID_ENDPOINT = exports.VALID_IDENTITY_ENDPOINT = exports.VALID_ID_SELECT = exports.VALID_NONCE = exports.OPENID_QUERY_PROPS = void 0;
/**
 * Query parameters that are allowed for be on the authentication request.
 */
exports.OPENID_QUERY_PROPS = [
    'openid.ns',
    'openid.mode',
    'openid.op_endpoint',
    'openid.claimed_id',
    'openid.identity',
    'openid.return_to',
    'openid.response_nonce',
    'openid.assoc_handle',
    'openid.signed',
    'openid.sig',
];
// All URLs required for this package.
exports.VALID_NONCE = 'http://specs.openid.net/auth/2.0';
exports.VALID_ID_SELECT = `${exports.VALID_NONCE}/identifier_select`;
exports.VALID_IDENTITY_ENDPOINT = 'https://steamcommunity.com/openid/id';
exports.VALID_OPENID_ENDPOINT = 'https://steamcommunity.com/openid/login';
exports.PLAYER_SUMMARY_URL = 'https://api.steampowered.com/ISteamUser/GetPlayerSummaries/v2';
exports.VALID_ASSOC_HANDLE = '1234567890';
exports.VALID_SIGNED_FIELD = 'signed,op_endpoint,claimed_id,identity,return_to,response_nonce,assoc_handle';
