// passport-steam-openid/dist/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SteamOpenIdErrorType = void 0;
/**
 * Enum used to distinguish between user-related errors.
 *
 * @enum SteamOpenIdErrorType
 */
var SteamOpenIdErrorType;
(function (SteamOpenIdErrorType) {
    /**
     * Invalid mode for authentication, should trigger redirect.
     */
    SteamOpenIdErrorType[SteamOpenIdErrorType["InvalidMode"] = 0] = "InvalidMode";
    /**
     * Query has a wrong parameter or has excess parameters.
     */
    SteamOpenIdErrorType[SteamOpenIdErrorType["InvalidQuery"] = 1] = "InvalidQuery";
    /**
     * Steam rejected our query.
     */
    SteamOpenIdErrorType[SteamOpenIdErrorType["Unauthorized"] = 2] = "Unauthorized";
    /**
     * SteamId is not valid.
     */
    SteamOpenIdErrorType[SteamOpenIdErrorType["InvalidSteamId"] = 3] = "InvalidSteamId";
    /**
     * Nonce has expired.
     */
    SteamOpenIdErrorType[SteamOpenIdErrorType["NonceExpired"] = 4] = "NonceExpired";
})(SteamOpenIdErrorType || (exports.SteamOpenIdErrorType = SteamOpenIdErrorType = {}));
